### PR TITLE
#166272101 Article and User validations

### PR DESCRIPTION
**Pull request by akinyeleolat**
#### What does this PR do?
This PR creates validations for user sign up

#### Description of Task to be completed?
- install validatorjs
- create the middleware folder structure 
- Set up validation rules for signup and login
- Set up validations rules for article creations
- write the test for user signup validations
- write helper function for displaying validation errors
- write helper functions for checking the validation

#### How should this be manually tested?
- run command "npm run dev"
- input "http://localhost:3000/api/v1/auth/signup" post route in postman
- fill in first name, last name, email, password, username
- click on send

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[166272101](https://www.pivotaltracker.com/story/show/166272101)

### CODE REVIEW

**Review by mosinmiloluwa01 — APPROVED** [review]

File: server/tests/controller/user.spec.js [hunk 1]
```diff
@@ -18,17 +18,6 @@ describe('User signup', () => {
       });
   });
```
Comment (Eazybee) [inline server/tests/controller/user.spec.js:20]: Why did you delete this test?

File: server/fixtures/userValidationData.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+export const emptyRequest = {};
+/** Invalid email */
+export const invalidEmail = {
+  email: 'akinye',
+  username: 'test',
+  lastName: 'test',
+  firstName: 'test',
+  password: 'test'
+};
+/** Invalid password */
```
Comment (Eazybee) [inline server/fixtures/userValidationData.js]: You don't need this comment here, your code is self-explanatory

File: server/fixtures/userValidationData.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
+/** Test cases for user validation */
```
Comment (Eazybee) [inline server/fixtures/userValidationData.js]: Please refactor this according to the TTL suggestion this morning. You can meet with @mosinmiloluwa01  for how to go about it if need be

File: server/helpers/utils.js [hunk 4]
```diff
@@ -66,4 +67,53 @@ export const createEllipsis = (text, length = 200) => {
...
+/**
+   * validate endpoint
+   * @param {Object} data - data to be validated
+   * @param {Object} rules - rules for validation
+   * @param {Object} res - response body
+   * @param {Object} next - call next function
+   * @returns {Boolean} true - if validation passes
+   * @returns {Object} error Object - if validation fails
+   */
+
```
Comment (Eazybee) [inline server/helpers/utils.js]: you might want to remove this empty line.  jsdocs and the function should not be separated

File: server/middlewares/articleValidation.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import { validate } from '<helpers>/utils';
+/** Article validation */
```
Comment (Eazybee) [inline server/middlewares/articleValidation.js]: you might want to add a new line here

File: server/middlewares/articleValidation.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+/**
+   * validate create article endpoint
+   * @param {Object} req - request body
+   * @param {Object} res - response body
+   * @param {Object} next - call next function
+   * @returns {Function} validate - call validate function
+   */
+
```
Comment (Eazybee) [inline server/middlewares/articleValidation.js]: please adjust this line.

File: server/helpers/utils.js [hunk 7]
```diff
@@ -67,3 +68,50 @@ export const createEllipsis = (text, length = 200) => {
   }
   return `${text.slice(0, length - 3).trim()}...`;
 };
+/**
```
Comment (Eazybee) [inline server/helpers/utils.js:72]: you might want to leave new line here

File: server/middlewares/authValidations.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  const data = {
+    firstName,
+    lastName,
+    email,
+    username,
+    password
+  };
+  const signUpRules = {
+    email: 'required|email',
+    username: 'required|string|alpha|min:6',
```
Comment (Eazybee) [inline server/middlewares/authValidations.js]: username is not required upon signup

File: server/routes/v1/user.js [hunk 9]
```diff
@@ -1,10 +1,11 @@
 import express from 'express';
 import signup from '<controllers>/signup';
 import existingUserCheck from '<middlewares>/existingUserChecker';
-import userValidations from '<middlewares>/userValidations';
+import { validateUserSignUp, validateUserLogin } from '<middlewares>/authValidations';
 
 const router = express.Router();
 
-router.post('/signup', userValidations, existingUserCheck, signup);
+router.post('/signup', validateUserSignUp, existingUserCheck, signup);
```
Comment (Eazybee) [inline server/routes/v1/user.js:8]: I feel validateUserSignUp and existingUserCheck should just be one middleware except you are planning to reuse it somewhere else.

File: server/tests/middlewares/loginValidation.spec.js [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        expect(res.status && res.body.status).to.be.equal(400);
+        expect(res).to.be.an('object');
+        expect(res.body)
+          .to.have.property('error');
+        expect(res.body.error.email).to.be.an('array').that.does.include('The email field is required.');
+        expect(res.body.error.password).to.be.an('array').that.does.include('The password field is required.');
+        done();
+      });
+  });
+  it('should return 400 if email is invalid', (done) => {
```
Comment (Eazybee) [inline server/tests/middlewares/loginValidation.spec.js:27]: you might want to leave a new line here for readability

File: server/tests/middlewares/loginValidation.spec.js [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    chai
+      .request(app)
+      .post('/api/v1/auth/login')
+      .send(getUserData({ email: 'testing' }))
+      .end((err, res) => {
+        expect(res.body.error.email).to.be.an('array').that.does.include('The email format is invalid.');
+        done();
+      });
+  });
+  it('should return 400 if password is less than 8 characters', (done) => {
```
Comment (Eazybee) [inline server/tests/middlewares/loginValidation.spec.js:38]: same here

File: server/tests/middlewares/signUpValidation.spec.js [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+          .to.have.property('error');
+        expect(res.body.error.email).to.be.an('array').that.does.include('The email field is required.');
+        expect(res.body.error.lastName).to.be.an('array').that.does.include('The lastName field is required.');
+        expect(res.body.error.firstName).to.be.an('array').that.does.include('The firstName field is required.');
+        expect(res.body.error.username).to.be.an('array').that.does.include('The username field is required.');
+        expect(res.body.error.password).to.be.an('array').that.does.include('The password field is required.');
+        done();
+      });
+  });
+  it('should return 400 if email is invalid', (done) => {
```
Comment (Eazybee) [inline server/tests/middlewares/signUpValidation.spec.js:29]: same here

File: server/tests/middlewares/createArticleValidation.spec.js [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        expect(res).to.be.an('object');
+        expect(res.body)
+          .to.have.property('error');
+        expect(res.body.error.title).to.be.an('array').that.does.include('The title field is required.');
+        expect(res.body.error.body).to.be.an('array').that.does.include('The body field is required.');
+        expect(res.body.error.categoryId).to.be.an('array').that.does.include('The categoryId field is required.');
+        done();
+      });
+  });
+  it('should return true if categoryId is a valid UUID', () => {
```
Comment (Eazybee) [inline server/tests/middlewares/createArticleValidation.spec.js:36]: you might want to leave a new line here for readability

**Review by Eazybee — CHANGES_REQUESTED** [review]

File: server/tests/controller/user.spec.js (hunk 1 above)
Comment (akinyeleolat) [reply server/tests/controller/user.spec.js:20]: It is a repetition

File: server/routes/v1/user.js (hunk 9 above)
Comment (akinyeleolat) [reply server/routes/v1/user.js:8]: It is a different function and will not obey SRP if that is done.

**Review by Eazybee — APPROVED** [review]

File: package.json [hunk 14]
```diff
@@ -8,15 +8,18 @@
     "lint": "eslint '**/*js'",
     "lint:fix": "prettier-eslint --write '**/*js' ",
     "start-dev": "nodemon ./server/app.js --exec babel-node",
+    "start": "babel-node ./server/app.js --exec",
```
Comment (benfluleck) [inline package.json]: Do not use `babel-node` to build production servers. Go and do some research on this

File: package.json [hunk 15]
```diff
@@ -29,19 +32,16 @@
...
   },
   "devDependencies": {
-    "@babel/plugin-transform-runtime": "^7.4.4",
-    "@babel/preset-env": "^7.4.5",
-    "@babel/register": "^7.4.4",
-    "babel-plugin-module-resolver": "^3.2.0",
     "chai": "^4.2.0",
     "chai-as-promised": "^7.1.1",
     "chai-http": "^4.3.0",
+    "chai-uuid": "^1.0.6",
```
Comment (benfluleck) [inline package.json:51]: Why are you installing this? Are you manually testing uuids?

File: server/fixtures/articleValidationData.js [hunk 16]
```diff
@@ -0,0 +1,17 @@
+
+export const emptyArticleData = {};
+
+export const invalidTitle = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.';
```
Comment (benfluleck) [inline server/fixtures/articleValidationData.js]: Why is this title invalid ?

File: server/helpers/utils.js [hunk 17]
```diff
@@ -67,3 +68,51 @@ export const createEllipsis = (text, length = 200) => {
...
+   * validate endpoint
+   * @param {Object} data - data to be validated
+   * @param {Object} rules - rules for validation
+   * @param {Object} res - response body
+   * @param {Object} next - call next function
+   * @returns {Boolean} true - if validation passes
+   * @returns {Object} error Object - if validation fails
+   */
+export const validate = (data, rules, res, next) => {
```
Comment (benfluleck) [inline server/helpers/utils.js]: Please use proper arguments names to represent the values
next is not a parameter for a helper function

File: server/middlewares/articleValidation.js [hunk 18]
```diff
@@ -0,0 +1,28 @@
+import { validate } from '<helpers>/utils';
```
Comment (benfluleck) [inline server/middlewares/articleValidation.js]: Change the folder path to `server/middlewares/validations/article `

File: server/fixtures/articleValidationData.js [hunk 19]
```diff
@@ -0,0 +1,17 @@
+
```
Comment (benfluleck) [inline server/fixtures/articleValidationData.js]: There should be no separate file for validations, all these should go into the article file

File: server/middlewares/articleValidation.js [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+  } = req.body;
+  const data = {
+    title,
+    body,
+    categoryId
+  };
+  const createArticleRules = {
+    title: 'required|string|max:100',
+    body: 'required|string',
+    categoryId: ['required', UUIDFormat]
```
Comment (benfluleck) [inline server/middlewares/articleValidation.js]: Why is the format for validating categoryId different from the others

Comment (benfluleck) [reply server/middlewares/articleValidation.js]: change UUIDFormat to uuidFormat

File: package.json (hunk 15 above)
Comment (akinyeleolat) [reply package.json:51]: the validation test with regex rules but 'chai-uuid' test the actual uuids.

**Review by benfluleck — CHANGES_REQUESTED** [review]
Review Added

File: server/fixtures/articleValidationData.js (hunk 16 above)
Comment (akinyeleolat) [reply server/fixtures/articleValidationData.js]: The article title was set to not more than 100 characters and this title character is more than 100 characters.

File: package.json (hunk 15 above)
Comment (benfluleck) [reply package.json:51]: Where are you testing uuid point me to a line...

File: server/middlewares/articleValidation.js (hunk 20 above)
Comment (akinyeleolat) [reply server/middlewares/articleValidation.js]: The array pattern is required when testing the regex pattern and not pipe character as stated in validatorjs documentation.

File: server/tests/middlewares/signUpValidation.spec.js [hunk 21]
```diff
@@ -0,0 +1,73 @@
+import chai from 'chai';
+import chaiHttp from 'chai-http';
+import chaiUUID from 'chai-uuid';
+import app from '<server>/app';
+import { getUserData, emptyUserData } from '<fixtures>/user';
+
+chai.use(chaiHttp);
+chai.use(chaiUUID);
```
Comment (benfluleck) [inline server/tests/middlewares/signUpValidation.spec.js]: What is this testing in this file ?

File: server/fixtures/articleValidationData.js (hunk 16 above)
Comment (benfluleck) [reply server/fixtures/articleValidationData.js]: yeah more like 300

File: server/fixtures/articleValidationData.js [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+export const emptyArticleData = {};
+
+export const invalidTitle = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.';
+
+export const invalidCategoryId = 'hdkdksks';
+
+export const goodArticleData = {
+  title: 'Generated Text',
+  body: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
```
Comment (benfluleck) [inline server/fixtures/articleValidationData.js]: You can re-use the text in that long articleTitle

File: server/tests/middlewares/createArticleValidation.spec.js [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+        expect(res.body)
+          .to.have.property('error');
+        expect(res.body.error.title).to.be.an('array').that.does.include('The title field is required.');
+        expect(res.body.error.body).to.be.an('array').that.does.include('The body field is required.');
+        expect(res.body.error.categoryId).to.be.an('array').that.does.include('The categoryId field is required.');
+        done();
+      });
+  });
+
+  it('should return true if categoryId is a valid UUID', () => {
```
Comment (akinyeleolat) [inline server/tests/middlewares/createArticleValidation.spec.js:36]: This test if the categoryId is a valid UUID

File: server/tests/middlewares/createArticleValidation.spec.js [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+    chai
+      .expect(goodArticleData.categoryId).to.be.a.uuid();
```
Comment (akinyeleolat) [inline server/tests/middlewares/createArticleValidation.spec.js:38]: This test if the goodArticleData categoryId is a valid UUID

File: package.json (hunk 15 above)
Comment (akinyeleolat) [reply package.json:51]: I just did with a comment

File: server/tests/middlewares/createArticleValidation.spec.js (hunk 23 above)
Comment (benfluleck) [reply server/tests/middlewares/createArticleValidation.spec.js:36]: I agree however I don't see `chai-uuid` being imported in this file. Can you please remove it and see if it still works.

File: server/tests/middlewares/createArticleValidation.spec.js (hunk 24 above)
Comment (benfluleck) [reply server/tests/middlewares/createArticleValidation.spec.js:38]: I agree however I don't see chai-uuid being imported in this file. Can you please remove it and see if it still works.

**Review by benfluleck — APPROVED** [review]
